### ISSUE TITLE
Enhancement: ban-types should allow generics

### ISSUE THREAD

**Issue by jcqvisser** [typescript-eslint/typescript-eslint#7090]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/ban-types/

### Description

I'd like to propose that some mechanism for banning certain uses of nested generics is added.

Something like:
```js
module.exports = {
  rules: {
    "@typescript-eslint/ban-types": [
      "error",
      {
        types: {
          "SomeGeneric<SomeOtherGeneric<*>>": "don't nest these two!"
        },
      }
    ],
  }
}
```

...Which would cause the following:
```ts
const a: SomeGeneric<SomeOtherGeneric<number>> = { value: { data: 1 } };
//            ^ Don't use `SomeGeneric<SomeOtherGeneric<*>>` as a type. don't nest these two! typeeslint[@typescript-eslint/ban-types](https://github.com/typescript-eslint/typescript-eslint/blob/v4.28.0/packages/eslint-plugin/docs/rules/ban-types.md)
```

My use case is to warn folks nesting optionals, eg:

```ts
interface OptionalWithDataAndStatus<TData> {
  data: TData;
  status: "success";
}
interface OptionalWithStatusWithoutData {
  status: "error" | "loading";
}
type Optional<TData> = OptionalWithDataAndStatus<TData> | OptionalWithStatusWithoutData;

const a: Optional<Optional<number>> = { status: 'success', data: { status: 'success', data: 1 } };
              ^ should show warning
```

I realize a similar thing can be accomplished by constraining the optional, eg:
```ts
interface OptionalWithDataAndStatus<TData> {
  data: TData;
  status: "success";
}
interface OptionalWithStatusWithoutData {
  status: "error" | "loading";
}
type Optional<TData> = TData extends OptionalWithDataAndStatus<any>
  ? never
  : TData extends OptionalWithStatusWithoutData
  ? never
  : OptionalWithDataAndStatus<TData> | OptionalWithStatusWithoutData;

const a: Optional<Optional<number>> = { status: 'success', data: 1 }
//                                      ^ Type 'string' is not assignable to type 'never'.
```

...though that is somewhat a moot point.

### Fail

```typescript
interface OptionalWithDataAndStatus<TData> {
  data: TData;
  status: "success";
}
interface OptionalWithStatusWithoutData {
  status: "error" | "loading";
}
type Optional<TData> = OptionalWithDataAndStatus<TData> | OptionalWithStatusWithoutData;

const a: Optional<Optional<number>> = { status: 'success', data: { status: 'success', data: 1 } };
              ^ should show warning
```

### Pass

```typescript
interface OptionalWithDataAndStatus<TData> {
  data: TData;
  status: "success";
}
interface OptionalWithStatusWithoutData {
  status: "error" | "loading";
}
type Optional<TData> = OptionalWithDataAndStatus<TData> | OptionalWithStatusWithoutData;

const a: Optional<number> = { status: 'success', data: 1 };
```

**Comment by bradzacher** [comment]
I think this usecases is a bit too niche for us to consider building into the rule.
Additionally it's not trivial to build such a feature using raw strings as you have to do be careful to ignore whitespace and comments etc when matching.

Your best bet would be to build a matcher using `no-restricted-syntax`:
https://typescript-eslint.io/linting/troubleshooting/#how-can-i-ban-specific-language-feature
[example](https://typescript-eslint.io/play/#ts=5.0.4&sourceType=module&showAST=es&code=PQKgUApgzgNglgOwC4AIEHsC0AnaTtwDGSEAJplAJ7ICGAHgFwoDaYKKARBNtuthwBo2KAN7D2HKBBgRifDkw4AVAMpLKABwgAlCADNuEBIQjMkmiADkaAWwgA6BLYgoAvCgDkK9HYDiR7iIPAF0UAD4UVXUtAAUabGcSbABJBCgkGmQ4GiQ4dARwyLULXQNcY1NzLWs7R2c3T28-AIJCEMFxTjsoKBoAcwgFTlJ81BGUJAALFynEPo5hAF8wYLAQYDAwKpcmiH8EQMIAHiUI9yUAbk3CfPSUGiZd-cOj3YB5Ke5n1qOEAFcbAAjbhhM6iFAANxoMD+ECYIhQpByDxQAEYUIsMRcgA)
